lc0/src/neural/network_blas.cc

```diff
@@ -0,0 +1,207 @@
+  size_t index = 0;
+  for (auto& plane : inputplanes) {
+    for (auto i : IterateBits(plane.mask)) {
+      input_data[index+i] = plane.value;
+    }
+    index += 64;
+  }
+  assert(index == input_data.size());
+
+  auto policy_data = std::vector<float>(weights_.ip_pol_b.size()); // get_num_output_policy()
+  auto  value_data = std::vector<float>(weights_.value.bn_means.size()*64); //NUM_VALUE_INPUT_PLANES*64
+
+  printf("Network::evaluate: input parsed, calling network...\n");
+  forwardPass(input_data, policy_data, value_data);
+  printf("Network forward pass complete, raw output:\n");
+  for (size_t i = 0; i < value_data.size(); i++)
+    printf("%g ", value_data[i]);
+  printf("\n");
+  for (size_t i = 0; i < policy_data.size(); i++)
+    printf("%g ", policy_data[i]);
+
+
+  // Get the moves
+  auto policy = softmax(policy_data);
+
+  // Now get the score
+  auto output = innerproduct(value_data, weights_.ip2_val_w, weights_.ip2_val_b);
+  assert(output.size() == 1);
+  auto value = output[0];
+
+  value = std::tanh(value);
+  printf("returning network evaluation\n");
+  return std::pair<float, std::vector<float>>(value, policy);
+}
+
+std::vector<float> BlasNetwork::winograd_transform_f(const std::vector<float>& f, const int outputs, const int channels) {
+  // F(2x2, 3x3) Winograd filter transformation
+  // transpose(G.dot(f).dot(G.transpose()))
+  // U matrix is transposed for better memory layout in SGEMM
+  auto U = std::vector<float>(WINOGRAD_TILE * outputs * channels);
+  auto G = std::array<float, WINOGRAD_TILE>{1.0,  0.0,  0.0,
+                                            0.5,  0.5,  0.5,
+                                            0.5, -0.5,  0.5,
+                                            0.0,  0.0,  1.0};
+  auto temp = std::array<float, 12>{};
+
+  for (auto o = 0; o < outputs; o++) {
+    for (auto c = 0; c < channels; c++) {
+      for (auto i = 0; i < 4; i++) {
+        for (auto j = 0; j < 3; j++) {
+          auto acc = 0.0f;
+          for (auto k = 0; k < 3; k++) {
+            acc += G[i*3 + k] * f[o*channels*9 + c*9 + k*3 + j];
+          }
+          temp[i*3 + j] = acc;
+        }
+      }
+
+      for (auto xi = 0; xi < 4; xi++) {
+        for (auto nu = 0; nu < 4; nu++) {
+          auto acc = 0.0f;
+          for (int k = 0; k < 3; k++) {
+            acc += temp[xi*3 + k] * G[nu*3 + k];
+          }
+          U[xi * (4 * outputs * channels)
+            + nu * (outputs * channels)
+            + c * outputs
+            + o] = acc;
+        }
+      }
+    }
+  }
+
+    return U;
+}
+
+std::vector<float> BlasNetwork::softmax(const std::vector<float>& inputs, float temperature) {
+  auto outputs = std::vector<float>(inputs.size());
+  auto alpha = *std::max_element(begin(inputs), begin(inputs) + outputs.size());
+  alpha /= temperature;
+  auto denom = 0.0f;
+  for (size_t i = 0; i < inputs.size(); i++) {
+    auto val = std::exp((inputs[i]/temperature) - alpha);
```

Comment:
Is it correct to subtract maximum probability from all elements?
For me it feels that {0; 0; 0.5} and {0.5; 0.5; 1} should return different results after normalization, but I don't understand how it may happen here.

Reply:
I dunno, again, ask gcp. This is how it's done in lczero and thus also Francois' branch at any rate